File: src/gzlib_ffi.erl
-module(gzlib_ffi).

-export([compress/1, compress/2, uncompress/1, crc32/1, crc32/2]).

compress(Data) ->
  try
    Z = zlib:open(),
    Bs = try
      zlib:deflateInit(Z),
      B = zlib:deflate(Z, Data, finish),
      zlib:deflateEnd(Z),
      B
    after
      zlib:close(Z)
    end,
    {ok, iolist_to_binary(Bs)}
  catch _:_ -> {error, nil} end.

compress(Data, Level) ->
  try
    Z = zlib:open(),
    Bs = try
      zlib:deflateInit(Z, Level),
      B = zlib:deflate(Z, Data, finish),
      zlib:deflateEnd(Z),
      B
    after
      zlib:close(Z)
    end,
    {ok, iolist_to_binary(Bs)}
  catch _:_ -> {error, nil} end.

uncompress(Data) ->
  try
    Z = zlib:open(),
    Bs = try
        zlib:inflateInit(Z),
        B = zlib:inflate(Z, Data),
        zlib:inflateEnd(Z),
        B
    after
        zlib:close(Z)
    end,
    {ok, iolist_to_binary(Bs)}
  catch _:_ -> {error, nil} end.

crc32(Data) ->
  try {ok, erlang:crc32(Data)}
  catch _:_ -> {error, nil} end.

crc32(Crc, Data) ->
  try {ok, erlang:crc32(Crc, Data)}
  catch _:_ -> {error, nil} end.

File: src/gzlib_ffi.mjs
import { deflateSync, inflateSync, crc32 as nodeCrc32 } from "node:zlib"
import { BitArray$BitArray, BitArray$BitArray$data, Result$Ok, Result$Error } from "./gleam.mjs"

export function compress(data, level = undefined) {
  try {
    if (level !== undefined && level < 0)
      return Result$Error(undefined)
    const buffer = deflateSync(BitArray$BitArray$data(data), { level })
    return Result$Ok(BitArray$BitArray(new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength)))
  } catch (_) {
    return Result$Error(undefined)
  }
}

export function uncompress(data) {
  try {
    const buffer = inflateSync(BitArray$BitArray$data(data))
    return Result$Ok(BitArray$BitArray(new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength)))
  } catch (_) {
    return Result$Error(undefined)
  }
}

export function crc32(data) {
  return crc32Continue(0, data)
} 

export function crc32Continue(crc, data) {
  try {
    if (crc < 0 || crc > 4294967295)
      return Result$Error(undefined)
    if (data.bitSize === 0)
      return Result$Ok(crc)
    return Result$Ok(nodeCrc32(BitArray$BitArray$data(data), crc))
  } catch (_) {
    return Result$Error(undefined)
  }
}
